go/vt/vtgate/planbuilder/operators/offset_planning.go

```diff
@@ -126,11 +130,31 @@ func addColumnsToInput(ctx *plancontext.PlanningContext, root Operator) Operator
 		return in, NoRewrite
 	}
 
+	// while we are out here walking the operator tree, if we find a UDF in an aggregation, we should fail
+	failUDFAggregation := func(in Operator, _ semantics.TableSet, _ bool) (Operator, *ApplyResult) {
+		aggrOp, ok := in.(*Aggregator)
+		if !ok {
+			return in, NoRewrite
+		}
+		for _, aggr := range aggrOp.Aggregations {
+			if aggr.OpCode == opcode.AggregateUDF {
+				// we don't support UDFs in aggregation if it's still above a route
+				panic(vterrors.VT03033(sqlparser.String(aggr.Original.Expr)))
+			}
+		}
+		return in, NoRewrite
+	}
+
+	visitor := func(in Operator, _ semantics.TableSet, isRoot bool) (Operator, *ApplyResult) {
+		out, res := addColumnsNeededByFilter(in, semantics.EmptyTableSet(), isRoot)
+		failUDFAggregation(in, semantics.EmptyTableSet(), isRoot)
+		return out, res
+	}
+
 	return TopDown(root, TableID, visitor, stopAtRoute)
 }
 
-// addColumnsToInput adds columns needed by an operator to its input.
-// This happens only when the filter expression can be retrieved as an offset from the underlying mysql.
+// pullDistinctFromUNION will pull out the distinct from a union operator
 func pullDistinctFromUNION(_ *plancontext.PlanningContext, root Operator) Operator {
```

Comment:
nit: Not sure why this function is named like this. Both DISTINCT and UNION are keywords, it would have made sense to capitalize both or neither versus this weird mix.